Cluster messages together when sent by the same sender in a row

nicholasngai opened the pull request:
<!--
Thanks for contributing to the project!
Please help us keep this project in good shape by going through this checklist.
Replace the empty checkboxes [ ] below with checked ones [X] as they are completed
Remember, you can preview this before saving it.
-->

<!-- You can remove this first section if you have contributed before -->

### First time contributor checklist:

- [x] I have read the [README](https://github.com/signalapp/Signal-Desktop/blob/master/README.md) and [Contributor Guidelines](https://github.com/signalapp/Signal-Desktop/blob/master/CONTRIBUTING.md)
- [x] I have signed the [Contributor Licence Agreement](https://signal.org/cla/)

### Contributor checklist:

- [x] My contribution is **not** related to translations. _Please submit translation changes via our [Signal Desktop Transifex project](https://www.transifex.com/signalapp/signal-desktop/)._
- [x] My commits are in nice logical chunks with [good commit messages](http://chris.beams.io/posts/git-commit/)
- [x] My changes are [rebased](https://medium.freecodecamp.org/git-rebase-and-the-golden-rule-explained-70715eccc372) on the latest [`development`](https://github.com/signalapp/Signal-Desktop/tree/development) branch
- [x] A `yarn ready` run passes successfully ([more about tests here](https://github.com/signalapp/Signal-Desktop/blob/master/CONTRIBUTING.md#tests))
- [x] My changes are ready to be shipped to users

### Description

<!--
Describe briefly what your pull request changes. Focus on the value provided to users.

Does it address any outstanding issues in this project?
  https://github.com/signalapp/Signal-Desktop/issues?utf8=%E2%9C%93&q=is%3Aissue
  Reference an issue with the hash symbol: "#222"
  If you're fixing it, use something like "Fixes #222"

Please write a summary of your test approach:
  - What kind of manual testing did you do?
  - Did you write any new tests?
  - What operating systems did you test with? (please use specific versions: http://whatsmyos.com/)
  - What other devices did you test with? (other Desktop devices, Android, Android Simulator, iOS, iOS Simulator)
-->

This adds the chaining together of messages sent by the same sender in a row, making it more clear when one person has sent a string of messages in a row. This is specifically useful in group chats, when it can be harder to distinguish the sender in many incoming messages from different senders.

Before:
![image](https://user-images.githubusercontent.com/36081376/84547578-b773c900-acb8-11ea-9eb8-68b98bd25a68.png)

After:
![image](https://user-images.githubusercontent.com/36081376/84585468-e5deca80-adc4-11ea-88b5-51bb2a254adc.png)


And likewise for incoming messages.

Manual testing was mainly accomplished through the style guide and through participation in an existing Signal group chat to make sure things remain distinct. No new tests were written, although some of the style guide was modified to incorporate displaying chained messages for the different additional content (images, quotes, previews, etc.). The operating systems tested are macOS and Debian GNU/Linux.

nicholasngai commented:
Whoops, actually, some of the border rounding for the image grid is still a little off. Will update later tonight.

scottnonnenberg-signal commented:
Wow, this is quite an endeavor. Be sure to look closely at how your phone does things, because we want Desktop to match that.

Next, the `collapseMetadata` prop is meant for bubbles which aren't the most recent in a given group of messages. You've probably noticed no timestamp when your phone collapses message bubbles?

Lastly, what is your logic for grouping? If you look at your phone, you'll note that it's based on whether they have the same visual timestamp. So they have to be sent pretty close to each other, not just be from the same sender.

nicholasngai commented:
Interesting. Looking at https://github.com/signalapp/Signal-iOS/blob/master/Signal/src/ViewControllers/ConversationView/ConversationViewModel.m#L1239, it seems like messages are always clustered together if they have the same sender, but whether the metadata is collapsed is based on whether they have the same visual timestamp (as well as not being a failed message or disappearing message). Will update to match.

nicholasngai commented:
Additionally, mobile adds an additional item into the timeline to separate days, so we can probably break up a cluster if the day is different, as well.

nicholasngai commented:
I updated the logic to collapse metadata when the timestamp strings are the same, but I’m not sure that exactly matches what is going on in iOS. For example, I have 2 clustered messages with the string `2 HR AGO` (sent about a minute apart), but iOS isn’t collapsing the metadata for the first message. Likewise, I have two messages with `FRI 11:08 PM` and `FRI 11:09 PM`, but the top message’s metadata is getting collapsed, so I’m not entirely sure what’s going on.

nicholasngai commented:
Hm, also, the timeline height is throwing me off. Because message metadata could go from uncollapsed to collapsed (say, when two messages are sent 30 seconds apart, so they sometimes have the same timeline text `1m` and sometimes have different timeline texts `1m` and `2m`), it would need to trigger a re-render for the height every 30 seconds as the two timestamp texts sometimes match and sometimes don’t match. There is potentially a similar problem with matching message delivery statuses, as well. Thoughts?

scottnonnenberg-signal commented:
@nicholasngai Ah, so you have figured out the most-difficult part of this: the fact that message height changes along with these clustering changes, and that requires recalculation of the vertical alignment of everything. Look for the `hasMessageHeightChanged` function and related functionality in the Timeline.

nicholasngai commented:
It looks like `hasMessageHeightChanged` determines if a message has changed based solely on the current and previous message data, though the message data doesn’t actually change for a message which goes from uncollapsed to collapsed just because time has moved forward. I’m tempted to just create a Redux action that forces a recalculation of the height by adding it to the `heightChangeMessageIds` array and firing it from the `Message` component whenever it updates itself based on whether it just went from collapsed to uncollapsed and vice-versa.

nicholasngai commented:
Okay, done.

nicholasngai commented:
I do have to say it’s a little odd watching messages automatically collapse and uncollapsed themselves as their visual timestamps independently change to match or un-match with the message following it. Is this behavior something we want to keep? Or is there a better solution?

scottnonnenberg-signal commented:
> I do have to say it’s a little odd watching messages automatically collapse and uncollapse

If you watch the mobile apps, you'll note that they do the same thing.

So, I see that there are still two incomplete checkboxes in the original template. For the first, for the lint-deps, you'll need to go update `ts/util/lint/exceptions.json` some of the line numbers may have changed. This should be pretty simple.

For the last checkmark, maybe you could talk through the testing you've done on this?

Lastly, are you considering applying to join the Signal Desktop team? https://jobs.lever.co/signal/6cbff26c-290a-4e74-a56f-78e9783f3f90

nicholasngai commented:
I just fixed the lint-deps exceptions file and updated the line numbers. I described the (sparse-ish) testing method near the bottom my original post, but I connected the development version of Signal Desktop to my normal account and have been using it regularly without any issues, including a relatively large group chat that tends to have many of the threads, so it at least seems to work empirically.

nicholasngai commented:
@scottnonnenberg-signal Is there anything else that is needed for this PR?

scottnonnenberg-signal commented:
@nicholasngai A whole lot of testing. Even before we go to public beta with it. You can help out by enlisting people you know to run your branch in development mode. It's easy to copy your data from Production/Beta installs to your development instance.

nicholasngai commented:
Interesting. After staring at Signal iOS, it looks like the collapsed state of bubbles is only updated when the conversation is loaded, and the last message is updated when a new message is sent so it collapses if need be. Currently, this PR just live collapses and un-collapses automatically, which can cause weird undulating effects on the entire timeline. Should we try to emulate the current functionality of iOS instead?

EvanHahn-Signal commented:
Sorry for the lack of updates here.

We've made significant changes since the last update to this pull request, as you can see from all the merge conflicts. We have some intention of implementing this feature and have started to build the infrastructure to make this possible. I think we're going to implement this ourselves, but we will refer to your hard work as a template.

I'm going to close this issue because we won't merge it as is, but again, thank you for your efforts.
